File: csharp/etl/ETL.cs
namespace Exercism
{
    using System.Collections.Generic;
    using System.Linq;

    public static class ETL
    {
        public static Dictionary<string, int> Transform(Dictionary<int, IList<string>> input)
        {
            return input.SelectMany(i => i.Value.Select(v => new KeyValuePair<string, int>(v.ToLowerInvariant(), i.Key)))
                        .OrderBy(k => k.Key)
                        .ToDictionary(kv => kv.Key, kv => kv.Value);
        }
    }
}

File: haskell/etl/ETL.hs
module ETL (transform) where

import Data.Map (Map, fromList, foldWithKey)
import Data.Char (toLower)

transform :: Map Int [String] -> Map String Int
transform input = fromList $ foldWithKey transform' [] input where	
	transform' key values acc = map (\v -> (map toLower v, key)) values ++ acc	